src/main/java/innohack/gem/service/MatchService.java

```diff
@@ -0,0 +1,229 @@
+package innohack.gem.service;
+
+import com.google.common.collect.Lists;
+import innohack.gem.dao.IGEMFileDao;
+import innohack.gem.dao.IGroupDao;
+import innohack.gem.entity.GEMFile;
+import innohack.gem.entity.rule.Group;
+import innohack.gem.entity.rule.rules.Rule;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class MatchService {
+  private static HashMap<String, Collection<Group>> matchedFileGroupTable =
+      new HashMap<String, Collection<Group>>();
+  private static HashMap<String, HashMap<Rule, Boolean>> matchedFileRuleTable =
+      new HashMap<String, HashMap<Rule, Boolean>>();
+
+  private List<GEMFile> filesWithConflictMatch = new ArrayList<GEMFile>();
+  private List<GEMFile> filesWithoutMatch = new ArrayList<GEMFile>();
+
+  @Autowired IGroupDao groupDao;
+  @Autowired IGEMFileDao gemFileDao;
+
+  public static HashMap<String, Collection<Group>> getMatchedFileGroupTable() {
+    return matchedFileGroupTable;
+  }
+
+  public static void setMatchedFileGroupTable(
+      HashMap<String, Collection<Group>> matchedFileGroupTable) {
+    MatchService.matchedFileGroupTable = matchedFileGroupTable;
+  }
+
+  public static HashMap<String, HashMap<Rule, Boolean>> getMatchedFileRuleTable() {
+    return matchedFileRuleTable;
+  }
+
+  public static void setMatchedFileRuleTable(
+      HashMap<String, HashMap<Rule, Boolean>> matchedFileRuleTable) {
+    MatchService.matchedFileRuleTable = matchedFileRuleTable;
+  }
+
+  public synchronized boolean checkMatching(Group group, GEMFile file) {
```

Comment:
About the synchronized method... things gets really tricky if we want to be concurrent safe... perhaps you wanted to protect the objects instead? (matchedFileRuleTable and matchedFileGroupTable) E.g. If a call is made for both methods `checkMatching` and `onUpdateFile`, then the concurrency seems to break.

A simple way is to use ConcurrentHashMaps or to synchronized across these objects... or we could even ignore it all together as this is probably a very low throughput program (i.e. not expecting many people to concurrently use it)

Reply:
I made some changes in PR #107.

- Changed HashMap to ConcurrentHashMap 
- Removed all synchronized method and changed all write method to private
- Added one synchronized method for calling from outside of MatchService class